server/map_rpc_server.go

```diff
@@ -354,66 +354,98 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 		if err != nil {
 			return err
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &trillian.SetMapLeavesResponse{MapRoot: newRoot}, nil
+}
 
-		hkv := make([]merkle.HashKeyValue, 0, len(req.Leaves))
-		for _, l := range req.Leaves {
-			if err := checkIndexSize(l.Index, hasher); err != nil {
-				return err
-			}
-			l.LeafHash = hasher.HashLeaf(mapID, l.Index, l.LeafValue)
+// writeRevision ensures the desired write revision is available and claim a lock on this revision
+// in this transaction so that competing writes transactinos for the next revision cannot all be
+// committed.
+func (t *TrillianMapServer) writeRevision(ctx context.Context, tree *trillian.Tree, tx storage.MapTreeTX, rev int64) (int64, error) {
+	writeRev, err := tx.WriteRevision(ctx)
+	if err != nil {
+		return 0, err
+	}
+	if rev != 0 && writeRev != rev {
+		return 0, status.Errorf(codes.FailedPrecondition, "can't write to revision %v", rev)
+	}
+	glog.V(2).Infof("%v: Writing at revision %v", tree.TreeId, writeRev)
+	return writeRev, nil
+}
 
-			if err = tx.Set(ctx, l.Index, l); err != nil {
-				return err
-			}
-			hkv = append(hkv, merkle.HashKeyValue{
-				HashedKey:   l.Index,
-				HashedValue: l.LeafHash,
-			})
+// writeLeaves updates the leaf values, but does not calculate nor update the Merkle tree.
+func (t *TrillianMapServer) writeLeaves(ctx context.Context, tree *trillian.Tree, hasher hashers.MapHasher, tx storage.MapTreeTX, leaves []*trillian.MapLeaf, rev int64) ([]merkle.HashKeyValue, error) {
+	hkv := make([]merkle.HashKeyValue, 0, len(leaves))
+	for _, l := range leaves {
+		if err := checkIndexSize(l.Index, hasher); err != nil {
+			return nil, err
 		}
+		l.LeafHash = hasher.HashLeaf(tree.TreeId, l.Index, l.LeafValue)
 
-		// Work around a performance issue when using the map in
-		// single-transaction mode by preloading all the nodes we know the
-		// sparse merkle writer is going to need.
-		if t.opts.UseSingleTransaction && t.opts.UseLargePreload {
-			if err := doPreload(ctx, tx, hasher.BitLen(), hkv); err != nil {
-				return err
-			}
+		if err := tx.Set(ctx, l.Index, l); err != nil {
+			return nil, err
 		}
+		hkv = append(hkv, merkle.HashKeyValue{
+			HashedKey:   l.Index,
+			HashedValue: l.LeafHash,
+		})
+	}
+	return hkv, nil
+}
 
-		if err = smtWriter.SetLeaves(ctx, hkv); err != nil {
-			return err
+// updateTree calculates the latest merkle values within the tree for the given leaf updates at
+// the specified revision. The new signed map root is committed to storage and returned.
+func (t *TrillianMapServer) updateTree(ctx context.Context, tree *trillian.Tree, hasher hashers.MapHasher, tx storage.MapTreeTX, hkv []merkle.HashKeyValue, metadata []byte, rev int64) (*trillian.SignedMapRoot, error) {
+	// Work around a performance issue when using the map in
+	// single-transaction mode by preloading all the nodes we know the
+	// sparse merkle writer is going to need.
```

Comment:
While you are here: **M**erkle

Reply:
Nothing like a refactoring to find out all the stuff that has bothered people quietly for a long time :-D